initial implementation

jlizen opened the pull request:
Sharing initial commit for design feedback. Please pick it to pieces. Glad to split into smaller commits if it'd be helpful. I figured it was short enough that seeing the whole picture might be useful.

Some specific questions:

## Crate name
Is `compute-heavy-future-executor` too long? Most crates are like 6 letters long. I was trying to make it clear what it does, but it's pretty klunky. Open to alternatives

## Strategy-setting API
I don't love the current ergonomics of initializing a strategy. We have all these calls like `initialize_block_in_place_strategy()` which either initialize the oncelock or panic.

I was considering something like a builder or an enum, but it felt ridiculous given that there is only really one step for these and then we stash any output in the once lock. And then they are more verbose to call.

Other ideas welcome.

## Custom executor
I wanted to put in an escape hatch to allow alternative async runtimes / customizing existing strategies with extra metrics or whatever / etc. But, the current form isn't particularly pleasant.

Ideally the caller would be able to implement the `ComputeHeavyFutureExecutor` trait themselves, but there were a few issues:
- We can't just accept a generic since we need to store the executor in the once lock, and we can't have generics in the static context
- We can't accept `Box<dyn ComputeHeavyFutureExecutor>` because the trait isn't object safe to its own generic bounds on its `execute(fut: F)` method

Instead I stored a closure inside of the `CustomExecutor` struct. That closure takes a future with `Any` output on the way in, and returns one with `Any` output. Because our `execute` call does have concrete typing, we can temporarily type erase the future's output before sending it into the closure, and then downcast it back on the flip side.

I think it's sound (please correct me, of course). Unfortunately we only are validating that the closure we get doesn't mangle the type, on initialization, not at compile time.

In addition to complexity, this also adds quite a few extra allocations due to all the added vtables and such. 

I'm open to alternative approaches. Or, if it's better just to drop this functionality from the library entirely, and force people to stay on the rails, I'm open to that too.

## Using `get_or_init()` inside `spawn_compute_heavy_future()`
Feedback on some dummy code in https://github.com/rustls/tokio-rustls/issues/94#issuecomment-2518177116 was that `get_or_init()` was overkill top call from inside the spawn handling.

As best as I can tell, using `get_or_init()` rather than matching against the runtime flavor every time this function is called, is strictly more efficient - since, it anyway starts off with `OnceCell::get()`, and then only calls the secondary logic if it was uninitialized - ref https://doc.rust-lang.org/beta/src/std/sync/once_lock.rs.html#387

This means that with `get_or_init()`, in case we need to use defaults, we have the single get branch every call, and the first call, the additional initialization logic.

Meanwhile with just a naked `get().unwrap_or_else()`, we still have the single get branch every call, and then the additional match branch against runtime flavor.

Probably a minor point either way, just calling out my reasoning so that it can be corrected as needed :)

## Use of log crate
It felt strange depending on `tracing` when tokio itself is optional. I was browsing other libraries and was seeing that many of them just use the simple `log` crate. Which has interop with tracing anyway.

Is that the right choice? Or are libraries just using `log` because they were written before `tracing` was widely used?

## Testing
Unit tests + doc tests succeed locally, with and without the tokio cfg flag enabled. I didn't want to overload this PR even further by setting some github actions, but I'll get to it!

arielb1 commented on src/custom_executor.rs:
```diff
@@ -0,0 +1,71 @@
+        F: Future<Output = O> + Send + 'static,
+        O: Send + 'static,
+    {
+        let wrapped_future = Box::pin(async move {
```
I have a feeling that using a channel is cleaner than using `Any` for the output type.

arielb1 commented on src/custom_executor.rs:
```diff
@@ -0,0 +1,71 @@
+    {
+        let wrapped_future = Box::pin(async move {
+            let res = fut.await;
+            Box::new(res) as Box<dyn Any + Send>
```
Not sure whether using `Any` or a channel is smarter here.

jlizen replied:
`Any` would no longer be agnostic to underlying caller's custom behavior right?

jlizen replied:
We discussed, the channel would let us erase the output type and replace with `()` at the cost of slightly more overhead from the channel. But then the input API would be nicer probably since less boxing by the caller.

Will poke around with this.

jlizen replied:
I have it working fine with the channel, but I was thinking more, and one use case I want to support is selective handling based on future type (for prioritization, etc). So totally type erasing the future's output type might not be ideal. Will poke around to see if there is a middle ground that preserves ergonomics.

jlizen replied:
Ehh, I decided not to scope creep on this. We can always add another strategy or a toggle on the existing one if this becomes an important use case.

arielb1 commented on src/lib.rs:
```diff
@@ -0,0 +1,550 @@
+    SecondaryTokioRuntime(SecondaryTokioRuntimeExecutor),
+}
+
+<<<<<<< Updated upstream
```
??? merge conflict marker

jlizen replied:
ah, whoops, thanks

arielb1 commented on src/lib.rs:
```diff
@@ -0,0 +1,550 @@
+                #[cfg(compute_heavy_executor_tokio)]
+                match tokio::runtime::Handle::current().runtime_flavor() {
+                    tokio::runtime::RuntimeFlavor::MultiThread => {
+                        log::info!("spawn_compute_heavy_future called without setting an explicit executor strategy, \
```
The `log` should probably be `trace` in production to avoid creating too much log spam.

arielb1 commented on src/lib.rs:
```diff
@@ -0,0 +1,550 @@
+    }
+}
+
+// Use nextest to run these, otherwise there will be race conditions against the strategy oncecell
```
Use doctests to ensure separate processes?

jlizen replied:
Oh, is that conventional? I can poke around at that.

I figured I can run them however i want (with nextest) in ci/cd anyway. But probably better to make it 'just work' locally.

jlizen replied:
We discussed, I'll just use integration tests.

arielb1 commented on src/secondary_tokio_runtime.rs:
```diff
@@ -0,0 +1,117 @@
+            )
+        });
+
+        match self.tx.try_send(Box::pin(background_future)) {
```
why are you using `try_send` instead of `send` here? You wait for a slot in the queue anyway.

jlizen replied:
Trying to avoid an extra task spawn. Is that not worth the added complexity?

jlizen replied:
Ah, I think I wrote this when I wasn't in an async context and moved it over. Will fix it.

arielb1 commented on src/lib.rs:
```diff
@@ -0,0 +1,550 @@
+    R: Send + 'static,
+{
+    let executor = COMPUTE_HEAVY_FUTURE_EXECUTOR_STRATEGY
+        .get_or_init(|| {
```
old complaint about `get_or_init`.

jlizen replied:
Did you see my discussion in the overview? Interested in your feedback on that.

jlizen replied:
Ah ok, I see your explanation below, I can change it back.

jlizen replied:
Just for clarity - the issue with this approach is, that if some initialization causes `spawn_compute_heavy_future()` before the caller does, then it prevents the caller from later setting one.

Anyway the cost of branching is insignificant given it is predictable. So better to avoid the potential sharp edge.

arielb1 commented:
> Is compute-heavy-future-executor too long? Most crates are like 6 letters long. I was trying to make it clear what it does, but it's pretty klunky. Open to alternatives

I don't think a long name now is a problem.

> I don't love the current ergonomics of initializing a strategy. We have all these calls like initialize_block_in_place_strategy() which either initialize the oncelock or panic.

I believe you should return a `Result` rather than panicking. A framework-like library might want to supply a strategy without panicking.

A builder might be nicer to find references for. For example, one that is used like `compute_heavy_future::global_strategy().initialize_block_in_place()` form.

> I think it's sound (please correct me, of course). 

If you don't use `unsafe` then it has to be sound. It does add a few extra allocations tho [I think 3 - one for the future, one for boxing the output, one for the waiter], but these should not be bad for futures that are actually "compute-heavy".

> Meanwhile with just a naked get().unwrap_or_else(), we still have the single get branch every call, and then the additional match branch against runtime flavor.

This should not affect performance significantly. In any interesting case both branches should be perfectly predictable.

jlizen commented on src/lib.rs:
```diff
@@ -0,0 +1,550 @@
+#[cfg(compute_heavy_executor_tokio)]
+pub fn initialize_secondary_tokio_runtime_strategy_and_config(
+    niceness: Option<i8>,
+    thread_count: Option<usize>,
```
Note: also add a customizable queue size for the background job queue, also concurrency limit on running tasks.

jlizen commented on src/secondary_tokio_runtime.rs:
```diff
@@ -0,0 +1,117 @@
+}
+
+impl SecondaryTokioRuntimeExecutor {
+    pub(crate) fn new(niceness: Option<i8>, thread_count: Option<usize>) -> Self {
```
Note: also add a customizable queue size for the background job queue, also concurrency limit on running tasks.

jlizen commented on src/lib.rs:
```diff
@@ -0,0 +1,550 @@
+        O: Send + 'static;
+}
+
+enum ExecutorStrategy {
```
Also look at adding cancellable behavior for anywhere that uses spawn to avoid zombie connection in case of partially i/o future.

jlizen commented on Cargo.toml:
```diff
@@ -14,11 +14,23 @@ categories = ["asynchronous"]
 # calling libraries can use the convention of `cfg(compute_heavy_executor)` to enable usage of this crate
-unexpected_cfgs = { level = "warn", check-cfg = ['cfg(compute_heavy_executor)', 'cfg(compute_heavy_executor_tokio)'] }
\ No newline at end of file
+unexpected_cfgs = { level = "warn", check-cfg = ['cfg(compute_heavy_executor)', 'cfg(compute_heavy_executor_tokio)'] }
```
Per discussion, probably we should switch these to feature flags since anyway caller will want to directly depend on `compute-heavy-future-executor` to enable the strategy. And then they can add a direct dependency on eg `tokio-rustls` with the feature enabled to flip it over to using it.

Only issue would be if there were multiple MVs of this crate.

jlizen commented on src/custom_executor.rs:
```diff
@@ -0,0 +1,71 @@
+
+        if let Err(_) = executor.execute(test_future).await {
+            panic!(
+                "CustomExecutor strategy initialized with a closure that \
```
Probably we should include that if the returned box of the future is dropped, the internal future should be cancelled as well, and check that at runtime. EG make sure the sender handle of injected oneshot is dropped.

jlizen commented:
@arielb1  - all comments should be addressed, thanks again for the close read

The main thing I want to call out is, I ended up adding a secondary `OnceLock` for the default strategy, that is lazily set and then subsequently ignored if a strategy is initialized afterwards.

I needed this because I started having lifetime issues on the `Default` impl for `ExecutorStrategyImpl` once I added in the possible Semaphore usage, even though it is always `None` when constructed as a default.

I figured it was probably fine since it still avoids the sharp edge of libraries calling `spawn_compute_heavy_future()` before the caller initializes. Open to alternatives.

Otherwise, I tried to break changes into commits, though some got a bit bigger. Here's the rundown:

### Changes since last review
- migrate from config flags to features
- Add `global_strategy_builder()` method
- Optional concurrency control for all executors
- All executors are cancellable 
    - I considered adding a knob to remove this behavior, but figured it was scope creep and cancellable is the right default behavior. I can open a github issue to see if anybody is interested in that later.
- Moved the custom executor over to using a oneshot channel rather than having `dyn Any`
   - I considered keeping some support for `Any` in case the closure wants to know the future's output type for prioritization or other selective handling. But that's scope creep, I can cut a github issue to see if anybody cares about having support for that via a new executor / new knob.
- Removed the default to `BlockInPlace` strategy for multi-threaded executor, there are too many footguns with it. Anyway the caller can choose it explicitly.
- Moved tests to integration tests rather than requiring `nextest`, it's awkward because it forces a flat directory structure but it's still better than doctests I feel
- Uses `COMPUTE_HEAVY_FUTURE_EXECUTOR_STRATEGY::get()` rather than `get_or_init()` inside `spawn_compute_heavy_future()` (though there is now a secondary oncelock for the default strategy as mentioned above)
- Various other small cleanup / tweaks

arielb1 commented on src/block_in_place.rs:
```diff
@@ -0,0 +1,47 @@
+
+impl BlockInPlaceExecutor {
+    pub(crate) fn new(max_concurrency: Option<usize>) -> Result<Self, Error> {
+        log::info!("initializing compute-heavy executor with block in place strategy, max concurrency: {:#?}", max_concurrency);
```
This looks overly log-heavy

arielb1 commented on src/lib.rs:
```diff
@@ -0,0 +1,571 @@
+/// ## Error
+/// Returns an error if the global strategy is already initialized.
+/// It can only be initialized once.
+pub fn global_strategy_builder() -> Result<GlobalStrategyBuilder, Error> {
```
Shouldn't you error just at the end rather than in this function?

arielb1 commented on src/lib.rs:
```diff
@@ -0,0 +1,571 @@
+{
+    let (mut tx, rx) = tokio::sync::oneshot::channel();
+    let wrapped_future = async {
+        select! {
```
should you use a biased rather than a random select here? https://docs.rs/tokio/latest/tokio/macro.select.html#fairness

jlizen replied:
Thanks, I was thinking it was biased by default.

jlizen commented:
Additional feedback from ariel - better to keep log messages on setting strategy inside a shared function where the once lock is actually set. Similarly only error there rather than sprinkled throughout constructor.

arielb1 commented on src/lib.rs:
```diff
@@ -0,0 +1,571 @@
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ExecutorStrategy {
```
Shouldn't this be `#[non_exhaustive]`?

arielb1 commented on src/lib.rs:
````diff
@@ -0,0 +1,571 @@
+    /// assert_eq!(res, 5);
+    /// # }
+    /// ```
+    #[cfg(feature = "secondary_tokio_runtime")]
````
Should this have `#[must_use = "doesn't do anything unless used"]`?

arielb1 commented on src/lib.rs:
```diff
@@ -0,0 +1,576 @@
+    }
+}
+
+#[must_use = "doesn't do anything unless used"]#[derive(Default)]
```
cargo fmt missing

jlizen replied:
thanks

arielb1 commented:
LGTM but plz rustfmt

jlizen commented:
Appreciate the review @arielb1 

I'll follow up with separate issues on:
- avoiding holding permits across i/o sleeps (maybe impl a wrapper future)
- support for propagating future type inside the custom executor, for prioritization purposes
- support for more variability in handling based on caller-specified 'likiness to block' / 'length of block', etc

rcoh commented on tests/block_in_place_strategy.rs:
```diff
@@ -0,0 +1,58 @@
+        );
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 10)]
+    async fn block_in_place_concurrency() {
+        initialize();
+
+        let start = std::time::Instant::now();
+
+        let mut futures = Vec::new();
+
+        for _ in 0..5 {
+            let future = async move { std::thread::sleep(Duration::from_millis(15)) };
+            // we need to spawn here since otherwise block in place will cancel other futures inside the same task,
+            // ref https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html
+            let future =
+                tokio::task::spawn(async move { execute_compute_heavy_future(future).await });
+            futures.push(future);
+        }
+
+        join_all(futures).await;
+
+        let elapsed_millis = start.elapsed().as_millis();
+        assert!(elapsed_millis < 50, "futures did not run concurrently");
+
+        assert!(elapsed_millis > 20, "futures exceeded max concurrency");
```
if there are 10 workers, I'm not sure if this actually tests that `execute_compute_heavy_future` is working? Seems like the number of futures would need to exceed the number of workers?

jlizen replied:
yeah, good point, let me update these tests

jlizen replied:
Specifically, I think we want 4 workers here, so that we can validate both that concurrency limit is firing, but also that the futures are playing nicely with worker threads blocking

rcoh commented:
My main feedback here is that I'm not sure about the mental model / use cases.

It seems like the strategy to use isn't necessarily application-global (or at least, there isn't a sensible default). Instead, if may make more sense to pick it based on the individual future? The properties of the future (or the number of futures) might dictate that choice.

Or perhaps put another way: I am a customer. How should I pick between these different options? Is there one that is the best?

jlizen commented:
Thanks for that @rcoh . Well stated.

I do have a few issues open that touch on this:
- https://github.com/jlizen/compute-heavy-future-executor/issues/5
- https://github.com/jlizen/compute-heavy-future-executor/issues/3
- https://github.com/jlizen/compute-heavy-future-executor/issues/4

I think we need some clearer recommendations around defaults. My knee jerk without profiling would be to guess that, probably spawn blocking + concurrency control is the right 'default behavior' to handle assorted calls to this by libraries. And then, if you know you have an issue with frequently blocking futures, the secondary tokio runtime might be better.

I want to validate those assumptions a bit more via profiling and then intend to flesh out the defaults as well as provide better explanations + recommendations in the module docs.

W/r/t to selective handling per future, I agree with that as well. I see two parts of this:
- Better 'turnkey' handling for libraries to more granularly define likeliness/length of blocking. I kept initial PR scope narrower to avoid scope creep, but I want to probe this and probably publish with at least two modes of operation (likely 'sometimes/briefly blocking' and 'frequently/extended blocking'. Needs more validation first though.
- I do think there is a place for exposing the specific shape of the input future into a custom closure-based executor as well (or perhaps a new turnkey strategy that allows injecting a selector between strategies based on input type, without requiring the full custom executor closure). Need to play around with that more.

rcoh commented on src/spawn_blocking.rs:
```diff
@@ -0,0 +1,38 @@
+    ComputeHeavyFutureExecutor,
+};
+
+pub(crate) struct SpawnBlockingExecutor {
+    concurrency_limit: ConcurrencyLimit,
+}
```
I wonder if this should actually hold onto a handle—you could initialize it from `Runtime::current` at creation time.

jlizen replied:
The benefit of that being, it would ensure that the spawn_blocking happens in the originating context instead of calling, in case of multiple runtimes? Or something else?

rcoh commented on src/spawn_blocking.rs:
```diff
@@ -0,0 +1,38 @@
+
+        let (wrapped_future, rx) = make_future_cancellable(fut);
+
+        if let Err(err) = tokio::task::spawn_blocking(move || {
+            tokio::runtime::Handle::current().block_on(wrapped_future)
+        })
```
does this actually work? I think once a bad future is in the runtime, it's going to potentially tie up a worker

jlizen replied:
Hmm, I think I found that approach from something alice posted somewhere, let me dig it up.

It's currently test e2e with a thread::sleep so it seems to work as expected?
https://github.com/jlizen/compute-heavy-future-executor/blob/main/tests/spawn_blocking_strategy.rs#L36

jlizen replied:
I think it was from here, which isn't particularly authoritative
https://stackoverflow.com/questions/76965631/how-do-i-spawn-possibly-blocking-async-tasks-in-tokio

Going to poke around this more closely, thanks for the callout

jlizen replied:
Ah, alice also discusses it in discord eg https://discord.com/channels/500028886025895936/500336333500448798/1189652644881514586

And here:
https://users.rust-lang.org/t/tokio-from-async-to-sync-and-back-to-async-block-on-vs-spawn-blocking/83438/9

But that doesn't address it being a potentially blocking future

jlizen replied:
To summarize offline discussion - the behavior is kind of surprising after doing some tracing of polls vs threads. Calling `Handle::block_on()` from inside `spawn_blocking()` just spawns a task to a regular worker thread for that runtime. Which makes it pretty useless for our purposes.

Which kind of makes sense, how else would tokio know how to schedule the sleeping futures without a runtime, and it doesn't want to lazily spin up a current thread runtime for every spawn blocking thread. At that point probably you're better off managing your own threadpool full of current thread runtimes or using a delegated secondary multithreaded runtime.

Anyway, PR coming to rip spawn blocking out and default to secondary tokio runtime in case the tokio feature is enabled. I think probably we need to tear our `block_in_place` as well, for similar reasons.

Then meanwhile we have our custom executor as an escape hatch for the 'pool of current thread runtimes' case, other async runtimes, etc.

jlizen replied:
I'm also wondering if we should have a (still async) `execute_sync()` function that accepts a sync input function, for the 'definitely this is blocking and not containing futures' segments of a compute heavy workload.

Since, for those, there is a simpler use case that might just involves `spawn_blocking` along with concurrency control.

Today, library authors can pick between `spawn_blocking()` on them (pretty opinionated, also tokio-specific) or just intermingling them with futures (no caller ability to slice them apart from futures that will need to sleep). Whereas ideally they could configure a strategy for handling sync inside of async, which gives the consumer escape hatch to avoid `spawn_blocking()` (perhaps prefer `block_in_place()`, perhaps sending to rayon, etc).

jlizen replied:
cut https://github.com/jlizen/compute-heavy-future-executor/issues/8
